src/client/pythonEnvironments/index.ts

```diff
@@ -2,150 +2,165 @@
+    // Build the stack of composite locators.
+    const [caching, disposable] = await createCachingLocator(ext, envInfoService, locators);
+    ext.disposables.push(disposable);
+    locators = caching;
 
-    return new ExtensionLocators(nonWorkspaceLocators, workspaceLocators);
+    return locators;
 }
 
-async function initNonWorkspaceLocators(): Promise<IDisposableLocator[]> {
-    const locatorFactories:(()=> Promise<IDisposableLocator>)[] = [
-        // Common locator factory goes here.
-        createGlobalVirtualEnvironmentLocator,
-        createPyenvLocator,
-    ];
-
+function createNonWorkspaceLocators(
+    ext: ExtensionState,
+): ILocator[] {
+    let locators: (ILocator & Partial<IDisposable>)[];
     if (getOSType() === OSType.Windows) {
         // Windows specific locators go here
-        locatorFactories.push(
-            createWindowsRegistryLocator,
-            createWindowsStoreLocator,
-        );
+        locators = [
+            new GlobalVirtualEnvironmentLocator(),
+            new PyenvLocator(),
+            new WindowsRegistryLocator(),
+            new WindowsStoreLocator(),
+        ];
     } else {
         // Linux/Mac locators go here
-        locatorFactories.push(
-            createPosixKnownPathsLocator,
-        );
+        locators = [
+            new GlobalVirtualEnvironmentLocator(),
+            new PyenvLocator(),
+            new PosixKnownPathsLocator(),
+        ];
     }
-
-    return Promise.all(locatorFactories.map((create) => create()));
+    const disposables = (locators.filter((d) => d.dispose !== undefined)) as IDisposable[];
+    ext.disposables.push(...disposables);
+    return locators;
 }
 
-function getWorkspaceFolders() {
-    const rootAdded = new vscode.EventEmitter<vscode.Uri>();
-    const rootRemoved = new vscode.EventEmitter<vscode.Uri>();
-    vscode.workspace.onDidChangeWorkspaceFolders((event) => {
+function watchRoots(args: WatchRootsArgs): IDisposable {
+    const { initRoot, addRoot, removeRoot } = args;
+
+    const folders = vscode.workspace.workspaceFolders;
+    if (folders) {
+        folders.map((f) => f.uri).forEach(initRoot);
+    }
+
+    return vscode.workspace.onDidChangeWorkspaceFolders((event) => {
         for (const root of event.removed) {
-            rootRemoved.fire(root.uri);
+            removeRoot(root.uri);
         }
         for (const root of event.added) {
-            rootAdded.fire(root.uri);
+            addRoot(root.uri);
         }
     });
-    const folders = vscode.workspace.workspaceFolders;
-    return {
-        roots: folders ? folders.map((f) => f.uri) : [],
-        onAdded: rootAdded.event,
-        onRemoved: rootRemoved.event,
-    };
+}
+
+function createWorkspaceLocator(
+    ext: ExtensionState,
+): WorkspaceLocators {
+    const locators = new WorkspaceLocators(
+        watchRoots,
+        [
+            // Add an ILocator factory func here for each kind of workspace-rooted locator.
```

Comment:
Can you also add the now exported `WorkspaceVirtualEnvironmentLocator` here?
```suggestion
            // Add an ILocator factory func here for each kind of workspace-rooted locator.
            (root: vscode.Uri) => [new WorkspaceVirtualEnvironmentLocator(root.fsPath)],
```

Reply:
I'd be glad to, but why didn't you do so in your PR?

Reply:
I had the notion of factory functions in my PR so `WorkspaceVirtualEnvironmentLocator` wasn't exported.
I could have connected the factory function instead but it was async and would change a lot of things. I figured those are going away anyways, so I didn't do it.